daemon/user/safelauncher.go

```diff
@@ -0,0 +1,98 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package user
+
+import (
+	"net/url"
+	"os/exec"
+
+	"fmt"
+
+	"github.com/godbus/dbus"
+)
+
+const (
+	safeLauncherIntrospectionXML = `
+	<interface name='com.canonical.SafeLauncher'>
+		<method name='OpenURL'>
+			<arg type='s' name='url' direction='in'/>
+		</method>
+	</interface>`
+)
+
+var (
+	allowedUrlSchemes = []string{"http", "https", "mailto"}
```

Comment:
we audit this already inside the core snap, before sending the call to the dbus ... i think it would be nice to agree to only do this auditing in one place (either at the start or at the end of the chain) so you do not need to change two places (core snap as well as snapd code) when adding or removing a protocol handler (i dont care which side we drop, but having a single place will prevent errors)